File: c/eu015.c
// project euler (projecteuler.net) problem 15
// solution by Kevin Retzke (retzkek@gmail.com)
// November 2012

#include <stdio.h>
#include <stdbool.h>
#include <gmp.h>

#include "pe.h"

unsigned long paths(unsigned int n)
{
	mpz_t fn, f2n, fSquared,  r;

	mpz_init(fn);
	mpz_fac_ui(fn,n);

	mpz_init(f2n);
	mpz_fac_ui(f2n,2*n);

	mpz_init(fSquared);
	mpz_mul(fSquared,fn,fn);

	mpz_init(r);
	mpz_cdiv_q(r, f2n, fSquared);

	if (!mpz_fits_ulong_p(r)) {
		printf("warning: result exceeds unsigned long\n");
	}
	return mpz_get_ui(r);
}

int main()
{
	pe_test_eq(paths(2), 6lu, "%lu");
	printf("%lu\n",paths(20));
}

File: go/eu015.go
// Project Euler problem 15
// Solution by Kevin Retzke
// December 2012
package main

import (
	"fmt"
	"math/big"
)

func factorial(n int) *big.Int {
	fac := new(big.Int)
	fac.MulRange(1, int64(n))
	return fac
}

func paths(n int) *big.Int {
	r := new(big.Int)
	fn := factorial(n)
	r.Div(factorial(2*n), fn.Mul(fn, fn))
	return r
}

func main() {
	fmt.Println(paths(20))
}
